### PR TITLE
Add IO to OpenEducation methodology

### PR DESCRIPTION
**Pull request by Mihnea0Firoiu**
# Prerequisite Checklist

- [x] Read the [contribution guidelines](https://github.com/cs-pub-ro/operating-systems/blob/main/CONTRIBUTING.md#pull-requests) regarding submitting new changes to the project;
- [x] Tested your changes against relevant architectures and platforms;
- [x] Updated relevant documentation (if needed).

## Description of changes

Align the IO course to the OpenEdu methodology.
Change the content of laboratories and change the "storyline" to:
* The Linux file-interface
  * file descriptors
  * file operations (I/O syscalls)
  * pipes and redirections
* Inter-process Communication
  * unix-sockets
  * network-sockets
  * client-server model
* Optimizations
  * I/O multiplexing
  * Async I/O
  * zero-copy

Besides these, in `quides/` you can find adjacent topics **not mandatory to finish the labs**, such as `file-descriptor table and open file table`, `file-mappings`, `kernel caching`, and `sendfile benchmarking`.

### CODE REVIEW

**Review by teodutu — CHANGES_REQUESTED** [review]
Check my comments to see some easy first fixes and also why the build fails. Then when it passes I'll make a more in-depth review.

- Move the task descriptions from the `reading/` files to each task's `README.md`
- Remove references to the common makefiles (`.../common/makefile/*.mk`). Copy their content to each makefile that includes them so the tasks don't depend on outside resources.

**Review by teodutu — COMMENTED** [review]
The build fails because `chapters/io/arena/drills/tasks/mini-shell/` is missing its `support/` folder which causes the Makefile symlink to point to an invalid path.

File: chapters/io/remote-io/drills/tasks/receive-challenges/support/Makefile
Comment (teodutu) [inline chapters/io/remote-io/drills/tasks/receive-challenges/support/Makefile:1]: Remove the `include` and copy-paste the contents of the included makefile here so the task doesn't have outside dependencies.

File: chapters/io/remote-io/drills/tasks/send-receive/support/receiver.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+import socket  # Exposes the Berkely Sockets API
+from os import getpid
+
+IP = "127.0.0.1"  # Loopback IP address (localhost)
+MAX_MESSAGE_LENGTH = 1024
+PORT = 5000  # Port to listen for data data on (non-privileged ports are > 1023)
```
Comment (teodutu) [inline chapters/io/remote-io/drills/tasks/send-receive/support/receiver.py:8]:
```suggestion
PORT = 5000  # Port to listen for data (non-privileged ports are > 1023)
```

**Review by teodutu — CHANGES_REQUESTED** [review]

File: chapters/io/io-internals/drills/tasks/buffering/support/libc_buffering.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
[the remark this answers is not in the document]
Comment (fabianpatras) [reply chapters/io/io-internals/drills/tasks/buffering/support/libc_buffering.c]: Why is this? (`//` vs `/* ... */`)

[the remark this answers is not in the document]
Comment (teodutu) [reply chapters/io/io-internals/drills/tasks/buffering/support/libc_buffering.c]: That's what checkpatch wants. I found this explanation, but I can't verify whether this is the main reason: https://github.com/ARM-software/tf-issues/issues/581#issuecomment-384960884

[the remark this answers is not in the document]
Comment (fabianpatras) [reply chapters/io/io-internals/drills/tasks/buffering/support/libc_buffering.c]: Yeah, that's what the checkpatch wants which is following this style:

```
C source: // SPDX-License-Identifier: <SPDX License Expression>
C header: /* SPDX-License-Identifier: <SPDX License Expression> */
[...]

If a specific tool cannot handle the standard comment style, then the appropriate comment mechanism which the tool accepts shall be used.
This is the reason for having the "/* */" style comment in C header files.
There was build breakage observed with generated .lds files where 'ld' failed to parse the C++ comment.
This has been fixed by now, but there are still older assembler tools which cannot handle C++ style comments.
```
Cited from https://github.com/torvalds/linux/blob/master/Documentation/process/license-rules.rst#license-identifier-syntax

[the remark this answers is not in the document]
Comment (teodutu) [reply chapters/io/io-internals/drills/tasks/buffering/support/libc_buffering.c]: But what you quoted explains why headers use C-style comments, not why source files use the C++-style.

**Comment by Alex-deVis** [comment]
@razvand, since I also made changes to the content itself, I think you should also review this.

File: chapters/io/file-descriptors/drills/tasks/my-cat/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# My `cat`
+
+Navigate to `file-descriptors/drills/tasks/my-cat/support/src` and checkout `my_cat.c`.
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/my-cat/README.md:3]:
```suggestion
Navigate to `chapters/io/file-descriptors/drills/tasks/my-cat/support/src` and checkout `my_cat.c`.
```
Use full paths just to be sure.

**Review by teodutu — CHANGES_REQUESTED** [review]
Why does this PR remove files from the `compute/` and `data/` chapters?

File: chapters/io/file-descriptors/drills/tasks/my-cat/solution/my_cat.c [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+		if (bytesRead == 0)
+			break;
+
+		totalBytesRead += bytesRead;
+		if (totalBytesRead == size)
+			break;
+	}
+
+	return totalBytesRead;
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/my-cat/solution/my_cat.c:39]: Convert these variables to `snake_case`. Do the same for the functions below.

File: chapters/io/ipc/drills/tasks/receive-challenges/README.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Receive Challenges
+
+Navigate to `chapters/io/ipc/drills/tasks/receive-challenges` and run `make` to generate the `support` directory.
```
Comment (teodutu) [inline chapters/io/ipc/drills/tasks/receive-challenges/README.md:3]: Remember to add the script, Makefile and `solution/` folder. And remove `support/`.

File: chapters/io/optimizations/drills/tasks/multiplexed-client-server/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
+# Multiplexed Client Server
+
+Navigate to `chapters/io/optimizations/drills/tasks/client-server-epoll` and run `make` to generate the `support` files.
```
Comment (teodutu) [inline chapters/io/optimizations/drills/tasks/multiplexed-client-server/README.md:3]:
```suggestion
Navigate to `chapters/io/optimizations/drills/tasks/multiplexed-client-server` and run `make` to generate the `support` files.
```

**Review by teodutu — CHANGES_REQUESTED** [review]

**Comment by Alex-deVis** [comment]
> Why does this PR remove files from the compute/ and data/ chapters?

Generic stuff such as removing `utils/sock` if unused and fixing links. Take a look for yourself:
```
$ git diff main $(git diff --name-only main . | egrep -v sock_util | grep compute)
```

**Comment by teodutu** [comment]
I think you forgot to add a file or messed up its name: https://github.com/cs-pub-ro/operating-systems/actions/runs/12047113855/job/33588983238#step:6:227

**Comment by Alex-deVis** [comment]
> I think you forgot to add a file or messed up its name: https://github.com/cs-pub-ro/operating-systems/actions/runs/12047113855/job/33588983238#step:6:227

I have no idea what happened - seems like the `shared-memory` task is not present in the current `copy-on-write` lab. Is this intended? I readded it to solve the issue.

**Review by teodutu — CHANGES_REQUESTED** [review]
The slides are still empty [1] because you forgot to add the `slides/` folders in each subchapter.

[1] https://cs-pub-ro.github.io/operating-systems/75/IO/

**Comment by Alex-deVis** [comment]
> The slides are still empty [1] because you forgot to add the slides/ folders in each subchapter.

It was a subtle metaphor for the need of I/O. Anyway, check again.

**Review by teodutu — APPROVED** [review]
Fuck that Java error.

**Pull request merged**